### PR TITLE
[OCTREE] Implementation of the iterator 'OctreeLeafNodeBreadthIterator'.

### PR DESCRIPTION
**Pull request by frozar**
**Warning:** this PR relies on #1983 

The purpose of this PR is to add a new iterator over leaf nodes of an octree. The existant leaf node iterator walks through an octree in the depth first way. The iterator I propose here walks through an octree in the breadth first way.

Technically, the implementation of `OctreeLeafNodeBreadthIterator` is really similar to `OctreeLeafNodeIterator` but it inherites from `OctreeBreadthFirstIterator`. So internally it uses a FIFO data structure.

The implementation is tested with the tests introduced in the PR #1983.

### CODE REVIEW

**Comment by SergioRAgostinho** [comment]
Question based on the assumption that all octree's leaf nodes are at the max depth. I haven't looked at the PR yet, just dumping the question based on the PR description.

Isn't the leaf iterator a special case of the FixedDepth iterator you're now implementing? Basically at max depth. Therefore, shouldn't it now derive from your new iterator type?

Opinions?

**Comment by frozar** [comment]
If you assume that all octree's leaf node are at the max depth, you're right, this iterator is just a special case of the FixedDepth iterator. But I don't want a so specific iterator.

Currently in PCL, you can assume the location of octree leaves because the octree are regular, initialised with the resolution of the deepest voxels. But in the case of irregular octree, I would like to get octree leaf iterator with the good behavior.

An irregular octree may seems like that:
capture du 2018-02-07 13-26-10
This image comes from the article "Octree-based region growing for point cloud segmentation". During the building, the refinement criterion depends if the points inside a current voxel seems like a plane or not. For such octrees, the iterators must be general purpose.

I already done the implementation of an octree with a refinement criterion like that. More PR will come, believe me ^^.

**Comment by SergioRAgostinho** [comment]
Sorry for the long delay. Let's discuss some ideas!

Is there a good reason for both LeafNode iterators coexist at the same time in the code base?

Assuming a well balanced tree, all nodes are populated, leaves all at the same level:
- `BreadthFirstLeafNode` it will take longer to initialize but iterating it after that will be quicker then the current implementation. 
- `DepthFirstLeafNode` will take less to initialize but iterating it has varying time.

Assuming the tree is no longer regular:
- `BreadthFirstLeafNode` - does this one still work? Does it means it visits the leaves which are at a shallower depth first in opposition to it's counterpart which goes deep first?

If one is gonna be called `LeafNodeBreadthIterator` than the other should also be renamed to `LeafNodeDeptthIterator`.

I see value in this. Opinions @PointCloudLibrary/maintainers

**Comment by frozar** [comment]
All the work I do about the iterators is to generalise thier behavior to irregular octrees. In the case of regular octree, as you said, there are pros/cons for both `BreadthFirstLeafNode` and `DepthFirstLeafNode`, so the user should be aware enough to choose one or another depend of his needs.

In the case of irregular octree, your remark is relevant: does it work correctly? Maybe a specific unit test about `LeafNodeBreadthIterator` should be added to ensure that point. Also, I must say that these iterators are not common. Usually, when iterator are present (in C++, in python, or whatever), the pitch says that the order in which you walk through a given container doesn't matter. But in some case, it matters and for some algorithm the way you walk through an octree makes a visible difference. So it is clear that in this PR, the iterator matter is the order of the traversal, and if I put a unit test, this would be the point to check (it was just to say).

Finally, you would like to rename `LeafNodeIterator` to `LeafNodeDepthIterator`, if you are OK to change the API, why not? Also, as everyone generally used the method `OctreeBase::leaf_begin()`/ `OctreeBase::leaf_end()` to initialise an iterator, maybe the renaming of these ones will have more impact on the client codes.

**Comment by SergioRAgostinho** [comment]
> Finally, you would like to rename LeafNodeIterator to LeafNodeDepthIterator, if you are OK to change the API, why not? Also, as everyone generally used the method OctreeBase::leaf_begin()/ OctreeBase::leaf_end() to initialise an iterator, maybe the renaming of these ones will have more impact on the client codes.

With some luck we get out of it with a simple
```cpp
typedef LeafNodeDepthIterator LeafNodeIterator;
```

**Comment by frozar** [comment]
Hello, this PR is ready for a review (finally).

The commits are currently splited, it would be easy this way to check what's going on here. Many commits regard some rename things. I wait for your feedback for thoses, and I'll squash them as you want.

Also, the API changes a bit, especially `leaf_begin()` is renamed `leaf_depth_begin()`. This change may impact other implementations. The rename of the iterator class itself is less import as I think everybody use the method `leaf_begin()` to deal with octree leaf nodes.

I also add a bunch of test over every iterator to check their behavior against an irregular octree. I try to build the tiniest representative irregular octree. I'll let you read the tests to check its understandability.

Finally, when I wrote the tests for Depth First iterators, it is clear now that they walk through an octree in reverse order. I would like to fix it in an other PR.

I stay tuned.

**Comment by frozar** [comment]
I'm waiting for your review guys...

**Comment by taketwo** [comment]
@SergioRAgostinho was taking care of Octree PRs before, so I assumed he'd be in position to review this one.

**Comment by SergioRAgostinho** [comment]
Which is a valid reasoning. However, I have very minimal time to look at this right now and this PR is not exactly a one liner.

My sincere apologies @frozar but as of now, I don't have time to dig into this.

File: octree/include/pcl/octree/octree_iterator.h [hunk 1]
```diff
@@ -756,6 +756,73 @@ namespace pcl
...
+         * \note step to next octree node
+         */
+        inline OctreeLeafNodeBreadthIterator
+        operator++ (int);
+
+        /** \brief *operator.
+         * \return pointer to the current octree leaf node
+         */
+        OctreeNode*
+        operator* () const;
```
Comment (taketwo) [inline octree/include/pcl/octree/octree_iterator.h]: Why do we need to have a custom `operator*`?

**Comment by taketwo** [comment]
Regarding the API change with the rename of `OctreeLeafNodeIterator`. We should not break existing code, give deprecation warning and transitioning grace period.

Unfortunately, the class is templated, so we can not simply typedef. However, I think it's reasonable to assume that nobody uses it directly, rather through the typedef given in the base octree class: 

```cpp
typedef OctreeLeafNodeIterator<OctreeT> LeafNodeIterator;
```

In other words, I assume that people write code that looks like:

```cpp
typename OctreeAdjacencyT::LeafNodeIterator leaf_itr = this->leaf_begin ()
```

rather than

```cpp
OctreeLeafNodeIterator<OctreeAdjacencyT> leaf_itr = this->leaf_begin ()
```

What we can do, therefore, is leave the old typedefs while adding new ones. That is, have:

```cpp
typedef OctreeLeafNodeDepthFirstIterator<OctreeT> LeafNodeIterator;
typedef OctreeLeafNodeDepthFirstIterator<OctreeT> LeafNodeDepthFirstIterator;
```

Of course, we can not attach a deprecation warning to the typedef. But we can (and should) attach one to the `leaf_begin` and `leaf_end` functions.

File: octree/include/pcl/octree/octree_iterator.h (hunk 1 above)
Comment (frozar) [reply octree/include/pcl/octree/octree_iterator.h]: Not needed.

**Comment by frozar** [comment]
@taketwo I'm glad that you give a try to review this PR.

For the rename and API stuff, I try as best as I understand your suggestion. If something goes wrong, just do an other proposition. So with the last commits, the API is not broken any more but it just evolve witht a deprecation message at compile time.

I think this PR is not really hard. There is a lot a of renaming and a new iterator, but nothing more. I stay tuned.

File: octree/include/pcl/octree/octree_base.h [hunk 2]
```diff
@@ -122,19 +123,36 @@ namespace pcl
         };
 
         // Octree leaf node iterators
-        typedef OctreeLeafNodeIterator<OctreeT> LeafNodeIterator;
-        typedef const OctreeLeafNodeIterator<OctreeT> ConstLeafNodeIterator;
+        // The previous deprecated names
```
Comment (taketwo) [inline octree/include/pcl/octree/octree_base.h]: Can you add a bit more guidance here? Something like:

```cpp
// LeafNodeIterator and ConstLeafNodeIterator are deprecated.
// Please use LeafNodeDepthFirstIterator and ConstLeafNodeDepthFirstIterator instead.
```

File: octree/include/pcl/octree/octree2buf_base.h [hunk 3]
```diff
@@ -248,10 +249,10 @@ namespace pcl
         const Iterator end() {return Iterator();};
 
         // Octree leaf node iterators
-        typedef OctreeLeafNodeIterator<OctreeT> LeafNodeIterator;
-        typedef const OctreeLeafNodeIterator<OctreeT> ConstLeafNodeIterator;
-        LeafNodeIterator leaf_begin(unsigned int max_depth_arg = 0) {return LeafNodeIterator(this, max_depth_arg);};
-        const LeafNodeIterator leaf_end() {return LeafNodeIterator();};
+        typedef OctreeLeafNodeDepthFirstIterator<OctreeT> LeafNodeDepthFirstIterator;
+        typedef const OctreeLeafNodeDepthFirstIterator<OctreeT> ConstLeafNodeDepthFirstIterator;
+        LeafNodeDepthFirstIterator leaf_begin(unsigned int max_depth_arg = 0) {return LeafNodeDepthFirstIterator(this, max_depth_arg);};
```
Comment (taketwo) [inline octree/include/pcl/octree/octree2buf_base.h]: Shouldn't we rename/deprecate here as well?

**Comment by frozar** [comment]
You see, we can do great job together.

If everything is ok, you can take a look to the next PR #2332 :wink:

File: test/octree/test_octree_iterator.cpp [hunk 4]
```diff
@@ -725,6 +767,290 @@ TEST_F (OctreeBaseIteratorsForLoopTest, FixedDepthIterator)
...
+      branch_count++;
+    }
+    else if (node->getNodeType () == pcl::octree::LEAF_NODE)
+    {
+      leaf_count++;
+    }
+    node_count++;
+  }
+
+  // Check the node_count, branch_count and leaf_count values
```
Comment (taketwo) [inline test/octree/test_octree_iterator.cpp]: Node count is not checked

**Comment by taketwo** [comment]
It feels like the Octree module is becoming the most thoroughly unit tested one in the library. Great job!

@SergioRAgostinho I'd say this is ready to merge, but in case you want to have a look (and will have time in foreseeable future), we can wait.

**Comment by SergioRAgostinho** [comment]
Getting closer to that day but still not there yet.

I'll wait for Fabien's update and then merge. From the quick glance I had at the commit messages I'm likely to squash this PR. Let me know in case you prefer to keep them separate, although given their granularity I would advise against.

**Comment by frozar** [comment]
Nice to hear from you. Nothing has been said against the `OctreeBaseWalkThroughIteratorsTest`, I like that 😄 

If the CI tests pass, it will be fine for me. You can squash this PR, I have no opinion about that. Or maybe the only constraint about the squash is to keep a pretty message that say there is a deprecation of some iterator (at least, I voluntarily forget the rename of iterator classes).

**Comment by taketwo** [comment]
Squashing all into one feels too crude. I propose to group the commits into three: a) all things related to the new iterator; b) renaming/deprecation of the existing iterator; c) additional tests.

**Comment by frozar** [comment]
@taketwo Ok, I'll do my best later ;)

**Comment by frozar** [comment]
Hello, here is my proposition for the squash stuff (this work was really hard), hope everybody will be fine with it.

**Comment by taketwo** [comment]
Looks good to me.

> this work was really hard

What was your procedure? I think in this case easiest was to soft-reset to master and then re-commit parts as needed.

**Comment by frozar** [comment]
What I did is many interactive rebase-s. The goal was to rebase this branch `octree_iterator_leaf` by firstly reordering the commits in the way you propose @taketwo and then to squash them.  
But as I have many renaming, un-renaming commit, it was not easy because I had many confilct to solve.

Even if I started from master, it was not obvious that I could done it quicker. Even if the different commits were focused on a given task, renaming in this case imply too many side effects. But it's done and well done, so everything is fine.

**Nota Bene:** the last tests I introduced assert that every available octree iterators have a correct behavior over an irregular octree!  
Next step: look at the PR #2332 which fixes the last odd behavior of 2 depth first iterators. (I cannot assign PR but I would give it to @taketwo 😉)

**Pull request merged**